File: Grating_Digital_Twin/mechanics/torque_model.m
function state = torque_model(torque,mechanical)
%TORQUE_MODEL Circular-shaft torque-to-torsion physical model.
%   theta = T*L/(G*J), J = pi*d^4/32.
%   All inputs use SI units. Default values are simulation assumptions and
%   must not be interpreted as patent dimensions or calibrated properties.
validateattributes(torque,{'numeric'},{'real','finite','vector'},mfilename,'torque');
required={'torsion_length','shear_modulus','shaft_diameter','grating_radius','grating_period'};
for k=1:numel(required)
    name=required{k};
    if ~isfield(mechanical,name)
        error('torque_model:MissingParameter','Missing mechanical parameter: %s.',name);
    end
    validateattributes(mechanical.(name),{'numeric'},{'scalar','real','finite','positive'},mfilename,name);
end
if ~isfield(mechanical,'phase_zero')||isempty(mechanical.phase_zero), mechanical.phase_zero=0; end
validateattributes(mechanical.phase_zero,{'numeric'},{'scalar','real','finite'},mfilename,'phase_zero');
if ~isfield(mechanical,'parameter_source')||isempty(mechanical.parameter_source)
    mechanical.parameter_source='simulation/default';
end
J=pi*mechanical.shaft_diameter^4/32;
theta_torsion=double(torque)*mechanical.torsion_length/(mechanical.shear_modulus*J);
relative_displacement=mechanical.grating_radius*theta_torsion;
optical_phase=2*pi*relative_displacement/mechanical.grating_period+mechanical.phase_zero;
state=struct('torque',double(torque),'polar_moment',J, ...
    'torsion_angle',theta_torsion,'relative_displacement',relative_displacement, ...
    'optical_phase',optical_phase,'parameter',mechanical, ...
    'parameter_source',char(mechanical.parameter_source), ...
    'equation','theta=T*L/(G*J), J=pi*d^4/32, dx=r*theta, phase=2*pi*dx/P');
end

File: Optical_Grating_Torque_DigitalTwin/mechanics/torque_model.py
"""Python reference implementation of the Stage-12 circular-shaft model."""
from __future__ import annotations

import math


def calculate_torque_state(
    *,
    torque: float,
    torsion_length: float,
    shear_modulus: float,
    shaft_diameter: float,
    grating_radius: float,
    grating_period: float,
    phase_zero: float = 0.0,
    parameter_source: str = "simulation/default",
) -> dict[str, float | str]:
    values = {
        "torsion_length": float(torsion_length),
        "shear_modulus": float(shear_modulus),
        "shaft_diameter": float(shaft_diameter),
        "grating_radius": float(grating_radius),
        "grating_period": float(grating_period),
    }
    if any(value <= 0.0 or not math.isfinite(value) for value in values.values()):
        raise ValueError("mechanical dimensions, modulus and grating period must be finite and positive")
    torque = float(torque)
    phase_zero = float(phase_zero)
    if not math.isfinite(torque) or not math.isfinite(phase_zero):
        raise ValueError("torque and phase_zero must be finite")
    polar_moment = math.pi * values["shaft_diameter"] ** 4 / 32.0
    torsion_angle = torque * values["torsion_length"] / (values["shear_modulus"] * polar_moment)
    relative_displacement = values["grating_radius"] * torsion_angle
    optical_phase = 2.0 * math.pi * relative_displacement / values["grating_period"] + phase_zero
    return {
        "torque": torque,
        "polar_moment": polar_moment,
        "torsion_angle": torsion_angle,
        "relative_displacement": relative_displacement,
        "optical_phase": optical_phase,
        "parameter_source": str(parameter_source),
    }
